### PR TITLE
Fix/#131 provide powerfactor mode

### PR DESCRIPTION
**Pull request by boltbeard**
This Pull Request also includes the feature of giving a timeseries input into eDisGo for generators (per technology and weather cell) and loads (per load type) in addition to adding power factor mode ('inductive' or 'capacitive')

### CODE REVIEW

File: edisgo/grid/components.py [hunk 1]
```diff
@@ -254,6 +257,65 @@ def peak_load(self):
...
+        """
+        if self._reactive_power_mode is None:
+            if isinstance(self.grid, MVGrid):
+                self._reactive_power_mode = self.grid.network.config[
+                    'reactive_power_mode']['mv_load']
+            elif isinstance(self.grid, LVGrid):
+                self._reactive_power_mode = self.grid.network.config[
+                    'reactive_power_mode']['lv_load']
+
+        return self._reactive_power_mode[1:-1]
```
Comment (birgits) [inline edisgo/grid/components.py]: Why is this necessary?

File: edisgo/grid/components.py [hunk 2]
```diff
@@ -254,6 +257,65 @@ def peak_load(self):
...
+        Get the sign reactive power based on the
+        :attr: `_reactive_power_mode`
+
+        Returns
+        -------
+        :obj: `int` : +1 or -1
+        """
+        comparestr = self.reactive_power_mode
+        comparestr = comparestr.lower()
+        if re.fullmatch('inductive', comparestr):
```
Comment (birgits) [inline edisgo/grid/components.py]: You can just do `if self.reactive_power_mode == 'inductive'`

File: edisgo/grid/components.py (hunk 1 above)
Comment (birgits) [reply edisgo/grid/components.py]: I mean [1:-1]

File: edisgo/grid/components.py [hunk 3]
```diff
@@ -184,11 +187,11 @@ def timeseries(self):
                 consumption = self.consumption[sector]
 
             if isinstance(self.grid, MVGrid):
-                q_factor = tan(acos(self.grid.network.config[
+                q_factor = self.q_sign * tan(acos(self.grid.network.config[
```
Comment (birgits) [inline edisgo/grid/components.py]: I would also add the attribute power_factor for loads to have it uniform to the generators.

**Comment by birgits** [comment]
@boltbeard looks good! Just some minor comments where I think the code could be made easier to understand.

File: edisgo/grid/components.py (hunk 1 above)
Comment (boltbeard) [reply edisgo/grid/components.py]: When 'inductive' and 'capacitive' are provided as strings in the config file, the first and the last character seems to be some non-readable characters. May be I could handle this better by directly influencing the parsing.

File: edisgo/grid/components.py (hunk 3 above)
Comment (boltbeard) [reply edisgo/grid/components.py]: I agree, I will add this in a later task though, it would be nice to have this. It wasn't there before and at the moment don't see any specific use for this, but maybe useful for the future.

File: edisgo/grid/components.py (hunk 2 above)
Comment (boltbeard) [reply edisgo/grid/components.py]: The idea was to use a regex match as a secure and clear check, I can change this.

File: edisgo/grid/components.py (hunk 1 above)
Comment (boltbeard) [reply edisgo/grid/components.py]: I just found out that this is caused by using the `'`..`'` symbols in the config files. I have removed these and also the [1:-1].

**Comment by boltbeard** [comment]
@birgits , changes made and tested.

File: edisgo/grid/network.py [hunk 4]
```diff
@@ -120,6 +120,11 @@ class EDisGo:
 
         Use 'other' if you don't want to explicitly provide every possible
         type.
+    timeseries_generation_reactive_power : :pandas:`pandas.DataFrame<dataframe>`, optional
```
Comment (birgits) [inline edisgo/grid/network.py]: Please describe how the dataframe has to look like.

Comment (birgits) [reply edisgo/grid/network.py]: And I thought it was gonna be normalized not kvar?

File: edisgo/grid/network.py [hunk 5]
```diff
@@ -1441,6 +1464,16 @@ class TimeSeries:
 
         Use 'other' if you don't want to explicitly provide every possible
         type. Default: None.
+    generation_reactive_power : :pandas: `pandasDataFrame<dataframe>`, optional
+        DataFrame with reactive power normalized with the nominal active power
+        rating or capacity which could be used to distribute the amount of
+        reactive power amongst all the available generators in the MV grid.
+        Time series can either be aggregated by technology type or by type
+        and weather cell ID. In the first case columns of the DataFrame are
+        'solar' and 'wind'; in the second case columns need to be a
```
Comment (birgits) [inline edisgo/grid/network.py]: What about dispatchable generators? What weather cell ID should they get?

File: edisgo/grid/network.py [hunk 6]
```diff
@@ -1526,6 +1561,22 @@ def generation_fluctuating(self):
...
+    def generation_reactive_power(self):
+        """
+        Get reactive power feedin time series normalized by active power
+        capacity.
+
+        Returns
+        -------
+        :pandas: `pandas.DataFrame<dataframe>`
+        """
+        return self._generation_reactive_power
```
Comment (birgits) [inline edisgo/grid/network.py]: Add loc[self.timeindex, :]

File: edisgo/grid/network.py [hunk 7]
```diff
@@ -1546,6 +1597,21 @@ def load(self):
...
+    @property
+    def load_reactive_power(self):
+        """
+        Get reactive power time series for load normalized by active power.
+
+        Returns
+        -------
+        :pandas: `pandas.DataFrame<dataframe>`
+        """
+        return self._generation_reactive_power
```
Comment (birgits) [inline edisgo/grid/network.py]: Add loc[self.timeindex, :]

File: edisgo/grid/components.py [hunk 8]
```diff
@@ -204,11 +207,49 @@ def timeseries(self):
...
+        return self._timeseries_reactive
+
+    @timeseries_reactive.setter
+    def timeseries_reactive(self, timeseries_reactive):
+        """
+        Setting the reactive power time series in kvar.
+
+        Attributes
+        ----------
+        timeseries_reactive : :pandas:`pandas.DataFrame<dataframe>`
```
Comment (birgits) [inline edisgo/grid/components.py]: Series or DataFrame?

Comment (birgits) [reply edisgo/grid/components.py]: At what point does the generator get its reactive power time series from the TimeSeries object?

File: edisgo/grid/components.py [hunk 9]
```diff
@@ -300,14 +436,19 @@ def timeseries(self):
...
         :attr:`_timeseries` is not :obj:`None`, it is returned. Otherwise,
         :meth:`timeseries` looks for time series of the according type of
-        technology in :class:`~.grid.network.TimeSeries`.
+        technology in :class:`~.grid.network.TimeSeries`. The reactive
+        power is also claculated in :attr:`_timeseries` depending upon the
+        :attr:`power_factor` and :attr:`reactive_power_mode`. The
+        :attr:`power_factor` determines the magnitude of the reactive power
+        based on the power factor and active power provided and the
+        :attr:`reactive_power_mode` determines if the reactive power is either
+        consumed (inductive behaviour) or provided (capacitive behaviour)
```
Comment (birgits) [inline edisgo/grid/components.py]: Only if no reactive power time series is given it is calculated.

File: edisgo/grid/components.py [hunk 10]
```diff
@@ -324,21 +465,87 @@ def timeseries(self):
...
+                self.power_factor = None
+                self.reactive_power_mode = None
+                self._timeseries_reactive = timeseries * self.nominal_capacity
+                return self._timeseries_reactive
+            except (KeyError, TypeError):
+                try:
+                    timeseries = \
+                        self.grid.network.timeseries.generation_reactive_power[
+                            'other'].to_frame('q')
+                    self.power_factor = 'not_applicable'
```
Comment (birgits) [inline edisgo/grid/components.py]: Why do you set power_factor to 'not_applicable' and in the first try to None?

File: edisgo/grid/components.py [hunk 11]
```diff
@@ -324,21 +465,87 @@ def timeseries(self):
...
+
+        Returns
+        -------
+        :pandas:`pandas.Series<series>`
+            series containing reactive power in kvar.
+        """
+        if self._timeseries_reactive is None:
+            try:
+                timeseries = \
+                    self.grid.network.timeseries.generation_reactive_power[
```
Comment (birgits) [inline edisgo/grid/components.py]: What if generation_reactive_power is a Multiindex DataFrame, will that work?

File: edisgo/grid/components.py (hunk 10 above)
Comment (boltbeard) [reply edisgo/grid/components.py]: The reason for this is to prevent the timeseries from being accidentally recalculated and changed some value, as if power_factor is set to None then it would re-calculate this from the assumption and get set to 0.9 or 0.95 as in the config file even though a given timeseries_reactive_power exists.

**Comment by boltbeard** [comment]
### Additional Task
- [ ] Make sure that no timeseries are saved in the components to reduce memory consumption > Keep using functions to calculate the timeseries instead and never save them.

File: edisgo/grid/components.py [hunk 12]
```diff
@@ -448,45 +723,125 @@ def timeseries(self):
...
-                    timeseries = self.grid.network.timeseries.\
+                    timeseries = self.grid.network.timeseries. \
                         generation_fluctuating[self.type].to_frame('p')
                 except KeyError:
                     logger.exception("No time series for type {} "
                                      "given.".format(self.type))
                     raise
 
-            timeseries['q'] = timeseries['p'] * tan(acos(self.power_factor))
             timeseries = timeseries * self.nominal_capacity
```
Comment (boltbeard) [inline edisgo/grid/components.py]: Make sure the subtract curtailment at this point, so that the correct reactive power is calculated from the `power_factor` also considers the reduced active power.
